### ISSUE TITLE
Make denote-open-or-create work better across subfolders

### ISSUE THREAD

**Issue by jonolo6** [protesilaos/denote#114]
When I am currently in a note that exists in a subfolder, if I invoke `denote-open-or-create` I only get options to search for notes in the subfolder. I believe this should search from the 'root' folder of that denote-project. I.e. that this should produce the same results as to how it currently works if I invoke the same command (`denote-open-or-create`) from a note in the root denote folder.

I.e. if I have a subfolder - lets call it 'subf' and I have some notes in this tree structure:

<note one>.org
<note two>.org
subf/<note three>.org
subf/<note four>.org

If I am in <note three>.org and invoke `denote-open-or-create` I get a search window that only produces results <note four>org. I would like the 'search' to also show <note one>.org and <note two>.org....

FYI - I am on DOOMEMACs and I use the default search (I'm a bit of a newbie to Emacs...)

**Comment by protesilaos** [comment]
Hello @jonolo6!

We have been making some changes to Denote to handle this use-case.  The idea is to treat your notes as a "project" and thus make all relevant searches start from the root directory, regardless of what subdirectory you are in.

Do you use Git?  If yes, try to initialise your `denote-directory` as a git project.  This will let Denote perform these project-wide searches.

Perhaps we can improve this so things "just work", though we are still in the development phase.

**Comment by nobiot** [comment]
@jonolo6 , @protesilaos , `denote-open-or-create` should already behave as you would like as of 6bb6da0 (20 November). I just tried the current main branch with Emacs with no configuration and it works on my end (the command can display files in sub-directories). 

@jonolo6  Could you please ensure that you have the latest commit of Denote? As you are on Doom Emacs, you would need to `doom sync` to properly effect the update, I think -- I'm not on Doom and it has been rather challenging for non-Doom users to support Doom users because of the differences. 

If you still do not see the behaviour change on the latest commit, let us know...

**Comment by jonolo6** [comment]
Yes this now seems to work as intended! Very much appreciated to see such quick responses! And yes - I'm using git for my files. In fact I'm using two silos as my employer does not allow cloud services for storing work notes so I sync with another service for my work and keep them separate. 

I have some more questions - let me know if you'd prefer me to create another issue here in Git for those:

### Question 1
I have the following set as I like to be asked about the subdirectory since I have my org-agenda-files pointing to the a subdirectory: `denote-prompts '(title keywords subdirectory)`. I'm adding it since perhaps this setting is impacting what I'm seeing below.
      
Is there a way to not have to enter the title twice when using `denote-open-or-create`. This is currently what happens:

1. I invoke `denote-open-or-create`
2. I get asked for a title that is being searched for as well. If I write out a title that's not matching another note and press enter then...
3. I again get asked for the title of the new note I want (Denote asks for`File title`)
4. Then keywords and subdirectory is being asked for, etc...

What I think should happen is that step 3 should be omitted as I already entered the full title in 2. Or at least somehow make it that way via an option via the `denote-prompts` or some other variable.

### Question 2
Regarding Silos. I find it a bit laborious to have to move between the silos to create and open notes in each. I understand notes should not link across silos, but I think it would be useful - at least via an option - to be able to open/search across the silos - independent of what project I'm currently in. Or ask for what silo to search for first as an option - similar to how I set the 'subfolders' option in `denote-prompts '(title keywords subdirectory)`. Perhaps like `denote-prompts '(title keywords **silo** subdirectory)`

What do you think?

**Comment by protesilaos** [comment]
> Is there a way to not have to enter the title twice when using `denote-open-or-create`.

The reason we have it like this is because the first prompt where you select a file uses pattern matching that may not be the exact title you want.  For example, I have the `orderless` package (I think Doom has it as well).  So I may write something like:

```
_journal test this ^2022
```

Which will match:

```
20221129T041717--this-is-a-test__journal.txt
```

To retrieve the first input, type `M-p` at the second prompt: it returns the last item from the relevant history.  Then edit it as you will.

> Regarding Silos. I find it a bit laborious to have to move between the silos to create and open notes in each. I understand notes should not link across silos, but I think it would be useful - at least via an option - to be able to open/search across the silos - independent of what project I'm currently in.

I have not tested this thoroughly, but I think it is possible to have an overarching project by specifying a silo whose main subdirectories are silos themselves.

What I did really quick:

* I have `~/Documents/notes` and `~/Documents/books`.  The latter is a silo with the relevant `.dir-locals`.
* I moved to `~/Documents` and created a `.dir-locals` there, declaring it a silo.
* When I am in the "notes" or "books", I only get relevant files from each of those.  But when I am in the Documents, I gain access to all of them.

Is this the idea?

The point with silos is to not know about the existence of each other.  Perhaps there is a way to integrate different silos via some customisation option, though this would probably require that we change how silos are defined.

**Comment by protesilaos** [comment]
Thank you @nobiot for your responses here!

About the specifics of the project integration, I will review everything holistically once all the code is in place.  For the time being, if I am in a subdirectory whose root directory is not controlled by Git, I do not gain access to all the files: I only see those of the subdirectory.  This has to do with how a project is recognised.  Again though, no worries about it right now, as I am not yet testing this in full.

**Comment by nobiot** [comment]
@jonolo6 , regarding the two silos issue, I think what @protesilaos suggests is worth trying. 

In addition, I use symlink like this. below In my case, without the second directory being a silo, but it may work well for your case.

```
~/Documents/notes  <= `denote-directory`
~/Documents/work/work-notes <= synced with work

~/Documents/notes/work -> symlink to ~/Documents/work/work-notes
```

**Comment by jonolo6** [comment]
Thx @nobiot @protesilaos

@protesilaos  - Understand the use case you explain re searching first and then entering the name. Makes full sense. I will try it out. I did not know about the `M-p` option and will see how that works for me.

Re the silos question:

I now tried this setup: 
~/denote/work <= here I had a .dir-locals file
~/denote/work/_agenda
~/denote/personal
~/denote/personal/_agenda 

And initially it seems fine. But work and personal are their individual git repos + with the silos I did find one issue: When being in a file in ~/denote/work/_agenda and do `denote-open-or-create` I only saw files in the .../work/_agenda location. When I did remove the .dir-locals under ~/denote/work it works as expected. I actually think I will stop using the silos for now and only keep work and personal in different git locations.

Removing the .dir-locals also works with org-journal that I have in another folder `~/org/journal`. So I can link into any work or private project should I want to from my journal...

Other than that I'm happy with the setup now. Thx for all your help!

**Comment by nobiot** [comment]
Hi jonolo6 

Just to quickly follow up on this:

>  When being in a file in ~/denote/work/_agenda and do denote-open-or-create I only saw files in the .../work/_agenda location.

This is very likely because the content of your `.dir-locals.el` is exactly as described in the manual:

```
;; Content of /home/nobiot/Documents/denote-test-local-dir/.dir-locals.el
((nil . ((denote-directory . default-directory))))
```

This means Emacs will set the value of `denote-directory`  to the current directory (that's what `default-directory` would mean) -- meaning, if you are in a sub-directory, that's set to the value of `denote-directory`. 

If I change the value to something like this -- the root of my silo, it works fine. See the screen shot below. I am showing the minibuffer selection when I call `denote-open-or-create` in a sub-directory of a silo.

```
;; Content of /home/nobiot/Documents/denote-test-local-dir/.dir-locals.el
((nil . ((denote-directory . "~/Documents/denote-test-local-dir"))))
```

![image](https://user-images.githubusercontent.com/12507865/204854514-f7d0ed6b-104c-4bc3-a299-488e6b9f4729.png)
